File: array/C++/dynamic_array.cc
/**
 * @file dynamic_array.cc
 * @author Hongyu Ding
 * @brief This is an implementation of dynamic array
 * @version 0.1
 * @date 2020-10-24
 * 
 * @copyright Copyright (c) 2020
 * 
 */

#include <iostream>

template <typename T>
class DArray {
    T* array;
    int capacity;
    int size;
public:
    DArray() : capacity{8}, size{0} {
        array = new T[capacity];
        clean();
    }
    DArray(T* a, int n) : capacity{n}, size{n} {
        array = new T[n];
        for (int i = 0; i < n; i ++) {
            array[i] = a[i];
        }
    }
    ~DArray() {
        delete[] array;
    }

    void clean() {
        for (int i = 0; i < size; i ++) {
            array[i] = 0;
        }
    }

    int getCapacity() const { return capacity; }
    int getSize() const { return size; }
    bool isEmpty() const { return size == 0; }

    T operator[](int i) const { return array[i]; }
    void set(int i, T value) {
        array[i] = value;
    }

    void add(T val) {
        if (size + 1 >= capacity) {
            if (capacity == 0) {
                capacity = 1;
            } else {
                capacity *= 2;
            }
            T* newArray = new T[capacity];
            for (int i = 0; i < size; i ++) {
                newArray[i] = array[i];
            }
            delete [] array;
            array = newArray;
        }
        array[size] = val;
        size ++;
    }

    T removeAt(int removeIndex) {
        T* newArray = new T[capacity];
        int index = 0;
        int removed = 0;
        for (int i = 0; i < size; i ++) {
            if (removeIndex != i) {
                newArray[index] = array[i];
                index ++;
            } else {
                removed = array[i];
            }
        }
        delete[] array;
        array = newArray;
        size --;
        return removed;
    }

    int indexOf(T val) {
        for (int i = 0; i < size; i ++) {
            if (array[i] == val) {
                return i;
            }
        }
        return -1;
    }

    bool contains(T val) {
        return indexOf(val) != -1;
    }

    void print(std::ostream& out) {
        if (isEmpty()) {
            out << "[]";
            return;
        }
        out << "[" << array[0];
        for (int i = 1; i < size; i ++) {
            out << ", " << array[i];
        }
        out << "]";
    }
};

int main() {
    DArray<int> array;
    array.add(2);
    array.add(5);
    array.add(6);
    array.add(1);
    array.add(-3);
    array.print(std::cout);
    array.removeAt(3);
    array.print(std::cout);
    return 0;
}

File: array&lists/Java/DynamicArray.java
/**
 * @File DynamicArray.java
 * @Author Weilan Tao
 * @Date 2020-12-02
 * @Description This is the implementation for Dynamic Array, Java version.
 * @Since version-1.0
 * @Copyright Copyright (c) 2020
 */


import java.util.Arrays;

public class DynamicArray<T> {
    private T[] array;
    private int capacity;
    private int size;

    public DynamicArray() {
        capacity = 8;
        size = 0;
        array = (T[]) new Object[capacity];

    }

    public int getCapacity() {
        return capacity;
    }

    public int getSize() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }


    public void clean() {
        int temp = size;
        for (int i = 0; i < temp; i++) {
            array[i] = null;
            size--;
        }
        capacity = 0;
    }

    public void setVal(int i, T t) {
        if (i < capacity) {
            array[i] = t;
        } else {
            System.out.println("index out of bounds");
        }
    }

    public void add(T t) {
        if (size < capacity) {
            array[size] = t;
            size++;
        } else {
            if (capacity == 0) {
                capacity = 1;
            } else {
                capacity *= 2;
            }
            T[] newarr = (T[]) new Object[capacity];
            System.arraycopy(array, 0, newarr, 0, size);
            newarr[size] = t;
            array = newarr;
            size++;
        }
    }

    public void removeAt(int index) {
        T[] removed = (T[]) new Object[capacity];
        size--;
        System.arraycopy(array, 0, removed, 0, index);
        System.arraycopy(array, index + 1, removed, index, capacity - index - 1);
        removed[capacity - 1] = null;
        array = removed;
    }

    public int indexOf(T t) {
        int index = -1;
        for (T data : array) {
            index++;
            if (data == (t)) {
                return index;
            }
        }
        return index;
    }

    public boolean contains(T val) {
        return indexOf(val) == -1;
    }

    public void print() {
        System.out.println(Arrays.toString(array));
    }

}

class TestDynamicArray {
    public static void main(String[] args) {
        DynamicArray<Integer> da = new DynamicArray<>();
        da.add(2);
        da.add(5);
        da.add(6);
        da.add(1);
        da.add(-3);
        da.add(-0);
        da.add(-9);
        da.add(100);
        da.add(20);
        da.print();
        System.out.println("getCapacity()..." + da.getCapacity());
        System.out.println("getSize()..." + da.getSize());
        System.out.println("indexOf()..." + da.indexOf(100));
        System.out.println("contains()..." + da.contains(101));

        da.removeAt(2);
        da.print();
        System.out.println("getCapacity()..." + da.getCapacity());
        System.out.println("getSize()..." + da.getSize());

        da.clean();
        System.out.println(da.isEmpty());
        da.print();

        da.add(19);
        da.print();
        System.out.println("getCapacity()..." + da.getCapacity());
        System.out.println(("getSize()..." + da.getSize()));

        da.setVal(2, 90);

    }
}
